### PR TITLE
Pass TimeStepperOptions to timesteppers

### PR DESCRIPTION
**Pull request by jwallwork23**
Also add some checks to `add_new_tracer`, as discussed in last Thetis meeting.

### CODE REVIEW

File: thetis/options.py [hunk 1]
```diff
@@ -726,16 +751,27 @@ def add_tracer_2d(self, label, name, filename, shortname=None, unit='-', source=
...
+        options.source = source
+        options.diffusivity = diffusivity
+        self.tracer_metadata[label] = AttrDict({
             'name': name,
             'shortname': shortname or name,
             'unit': unit,
             'filename': filename,
-            'source': source,  # TODO: Validate input
-        }
+            'options': options,
```
Comment (tkarna) [inline thetis/options.py:773]: So now we can access tracer options like this
```python
options.tracer_metadata[label].name
options.tracer_metadata[label].options.diffusivity
```
maybe this would be more logical
```python
options.tracer[label].metadata.name
options.tracer[label].diffusivity
```
i.e. the fields are not inside the metadata dict?

File: thetis/solver2d.py [hunk 2]
```diff
@@ -331,6 +331,15 @@ def add_new_field(self, function, label, name, filename, shortname=None, unit='-
...
         """
+        assert isinstance(function, Function)
+        assert isinstance(label, str)
+        assert isinstance(name, str)
+        assert isinstance(filename, str)
+        assert shortname is None or isinstance(shortname, str)
+        assert isinstance(unit, str)
+        assert preproc_func is None or callable(preproc_func)
+        if label in self.fields:
+            print_output(f"Field '{label}' already exists. It will be overwritten.")
```
[the remark this answers is not in the document]
Comment (tkarna) [reply thetis/solver2d.py:342]: While at it could you please add the equivalent to solver 3d as well?

**Review by tkarna — COMMENTED** [review]
Thanks, this is neat! Only have one comment on the tracer options API

File: thetis/solver2d.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jwallwork23) [reply thetis/solver2d.py:342]: I did initially put it in as an assertion, but some of the tests crashed because they initialise twice (I think without meaning to). Happy to switch it and then modify those tests. Can do for 3D, too.

File: thetis/options.py (hunk 1 above)
Comment (jwallwork23) [reply thetis/options.py:773]: Good idea, makes sense

File: thetis/solver2d.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jwallwork23) [reply thetis/solver2d.py]: Got it, the reason it was already in there is because the default 'tracer_2d' is hard-coded into `field_metadata`. Have removed it and made sure that it isn't in there when a solver object is created.

File: thetis/solver2d.py [hunk 3]
```diff
@@ -136,6 +136,8 @@ def __init__(self, mesh2d, bathymetry_2d, options=None):
 
         self.bnd_functions = {'shallow_water': {}, 'tracer': {}, 'sediment': {}}
 
+        if 'tracer_2d' in field_metadata:
+            field_metadata.pop('tracer_2d')
```
Comment (tkarna) [inline thetis/solver2d.py:140]: why is this needed?

File: thetis/solver2d.py [hunk 4]
```diff
@@ -374,28 +384,22 @@ def create_equations(self):
...
-                self.add_new_field(Function(self.function_spaces.Q_2d, name='tracer_2d'),
-                                   'tracer_2d',
-                                   md['name'],
-                                   md['filename'],
-                                   shortname=md['shortname'],
-                                   unit=md['unit'])
+            if self.options.tracer == {}:
+                self.options.add_tracer_2d('tracer_2d', 'Depth averaged tracer', 'Tracer2d', 'Tracer', '-',
+                                           source=self.options.tracer_source_2d,
+                                           diffusivity=self.options.horizontal_diffusivity)
```
Comment (tkarna) [inline thetis/solver2d.py:390]: I guess this is for backwards compatibility. Should we deprecate this code path in the future? That is, use always has to define tracers via `options.add_tracer_2d`? That would imply removing `options.solve_tracer`.

File: thetis/solver2d.py [hunk 5]
```diff
@@ -828,7 +776,7 @@ def print_state(self, cputime):
         :arg float cputime: Measured CPU time
         """
         if self.options.tracer_only:
-            for l in self.options.tracer_metadata:
+            for l in self.options.tracer:
```
Comment (tkarna) [inline thetis/solver2d.py:779]: `l` -> `label`? I think PEP8 suggests not to use variable name `l`

**Review by tkarna — APPROVED** [review]
Looks good! Only minor comments

File: thetis/solver2d.py (hunk 4 above)
Comment (jwallwork23) [reply thetis/solver2d.py:390]: Sure. We would then also be able to drop `options.horizontal_diffusivity` and `options.tracer_source_2d`. I'll add in a deprecation warning.

File: thetis/solver2d.py (hunk 3 above)
Comment (jwallwork23) [reply thetis/solver2d.py:140]: It is needed in the tests which do convergence studies. They create a sequence of solver objects and solve on each one. If this isn't included then the 2nd solver object tries to add a tracer field that already exists when it is instantiated.

Comment (tkarna) [reply thetis/solver2d.py:140]: Ok, so this could plausibly be removed when we deprecate `options.solve_tracer`. We still might need a workaround in the test but it's better to have it there rather than the solver object.

**Review by tkarna — APPROVED** [review]
